### PR TITLE
[FEATURE] Fluent `DataAsset` `batch_metadata` config variables

### PR DESCRIPTION
**Pull request by NathanFarmer**
Changes proposed in this pull request:
- GREAT-1729
- Add variable substitution for `batch_metadata` that references config variables

### Definition of Done
- [X] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/docs/contributing/style_guides/code_style)
- [X] I have performed a [self-review](https://docs.greatexpectations.io/docs/contributing/contributing_checklist) of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [X] I have added [unit tests](https://docs.greatexpectations.io/docs/contributing/contributing_test#writing-unit-and-integration-tests) where applicable and made sure that new and existing tests are passing.
- [X] I have run any local integration tests and made sure that nothing is broken.

### CODE REVIEW

File: great_expectations/datasource/fluent/interfaces.py [hunk 1]
```diff
@@ -241,6 +238,25 @@ def _validate_batch_request(self, batch_request: BatchRequest) -> None:
...
+        return set(options.keys()).issubset(set(self.batch_request_options))
+
+    def _get_batch_metadata_from_batch_request(
+        self, batch_request: BatchRequest
+    ) -> BatchMetadata:
+        """Performs config variable substitution and populates batch request options for
+        Batch.metadata at runtime.
+        """
+        batch_metadata = copy.deepcopy(self.batch_metadata)
+        config_variables = self._datasource._data_context.config_variables
```
Comment (NathanFarmer) [inline great_expectations/datasource/fluent/interfaces.py]: This means `_data_context` really is needed on all `Datasource`s.

File: tests/datasource/fluent/test_spark_datasource.py [hunk 2]
```diff
@@ -53,3 +72,48 @@ def test_dataframe_asset(
             for asset in datasource.assets.values()
         ]
     )
+
+
+@pytest.mark.xfail(
+    strict=True,
+    reason="this will fail until we add batch_metadata to Spark add_*_asset methods",
+)
```
Comment (NathanFarmer) [inline tests/datasource/fluent/test_spark_datasource.py:80]: @billdirks your PR for Spark `batch_metadata`, should cause you to remove this xfail.

File: great_expectations/datasource/fluent/pandas_datasource.pyi [hunk 3]
```diff
@@ -143,12 +143,14 @@ class PandasDatasource(_PandasDatasource):
     type: Literal["pandas"]
     assets: Dict[str, _PandasDataAsset]
     def test_connection(self, test_assets: bool = ...) -> None: ...
-    def _get_validator(self, asset: _PandasDataAssetT) -> Validator: ...
+    @staticmethod
+    def _validate_asset_name(self, asset_name: Optional[str] = ...) -> str: ...
+    def _get_validator(self, asset: _PandasDataAsset) -> Validator: ...
```
Comment (Kilo59) [inline great_expectations/datasource/fluent/pandas_datasource.pyi]: Why do these need to be part of the stub?
Are they being called from outside the class?

File: great_expectations/datasource/fluent/interfaces.py [hunk 4]
```diff
@@ -241,6 +238,25 @@ def _validate_batch_request(self, batch_request: BatchRequest) -> None:
...
+        batch_metadata = _ConfigurationSubstitutor().substitute_all_config_variables(
+            data=batch_metadata, replace_variables_dict=config_variables
+        )
```
Comment (Kilo59) [inline great_expectations/datasource/fluent/interfaces.py:252]: Does this also automatically prevent these values from getting serialized to config?

File: great_expectations/data_context/data_context/abstract_data_context.py [hunk 5]
```diff
@@ -1455,12 +1456,12 @@ def get_datasource(
         substituted_config = self.config_provider.substitute_config(raw_config_dict)
 
         # Instantiate the datasource and add to our in-memory cache of datasources, this does not persist:
-        datasource_config = datasourceConfigSchema.load(substituted_config)
```
Comment (NathanFarmer) [inline great_expectations/data_context/data_context/abstract_data_context.py:1458]: This variable was unused.

File: great_expectations/datasource/fluent/interfaces.py (hunk 4 above)
Comment (NathanFarmer) [reply great_expectations/datasource/fluent/interfaces.py:252]: Yes, the variables are never substituted on the `DataAsset` itself. This method only gets called in `get_batch_list_from_batch_request` and then the `BatchMetadata` is attached to each `Batch.metadata`.

File: great_expectations/datasource/fluent/interfaces.py [hunk 6]
```diff
@@ -462,6 +478,7 @@ def get_asset(self, asset_name: str) -> _DataAssetT:
         """Returns the DataAsset referred to by name"""
         # This default implementation will be used if protocol is inherited
         try:
+            self.assets[asset_name]._datasource = self
```
Comment (NathanFarmer) [inline great_expectations/datasource/fluent/interfaces.py:477]: The `get_datasource` and `get_asset` methods need to propagate updated `DataContext` -> `Datasource` -> `DataAsset` attributes. Otherwise you end up with missing `_data_context`s that are needed for config variable substitution and `get_validator` calls.

File: great_expectations/datasource/fluent/pandas_datasource.pyi (hunk 3 above)
Comment (NathanFarmer) [reply great_expectations/datasource/fluent/pandas_datasource.pyi]: @Kilo59 I forgot to follow up with you about this. I think they can actually be removed from the stubs (I removed them). It just results in the IDE complaining about "Unresolved attribute", but I think the stubs should be reserved for public methods.

File: great_expectations/data_context/data_context/abstract_data_context.py [hunk 7]
```diff
@@ -1443,6 +1443,7 @@ def get_datasource(
             )
 
         if datasource_name in self._cached_datasources:
+            self._cached_datasources[datasource_name]._data_context = self
```
Comment (NathanFarmer) [inline great_expectations/data_context/data_context/abstract_data_context.py:1446]: The `get_datasource` and `get_asset` methods need to propagate updated `DataContext` -> `Datasource` -> `DataAsset` attributes. Otherwise you end up with missing `_data_context`s that are needed for config variable substitution and `get_validator` calls.

File: great_expectations/data_context/data_context/abstract_data_context.py (hunk 5 above)
Comment (Kilo59) [reply great_expectations/data_context/data_context/abstract_data_context.py:1458]: I would think the linter would have flagged this 🤷

File: great_expectations/datasource/fluent/interfaces.py (hunk 6 above)
Comment (Kilo59) [reply great_expectations/datasource/fluent/interfaces.py:477]: You might have just fixed an edge case I found with my `ConfigStr` logic 😍 
It was happening because of a missing `_data_contexts.config_provider` which I didn't understand at the time.

**Review by Kilo59 — APPROVED** [review]
🙌

File: tests/datasource/fluent/test_spark_datasource.py (hunk 2 above)
Comment (billdirks) [reply tests/datasource/fluent/test_spark_datasource.py:80]: This actually still fails because the `DataFrameAsset` has not implemented `build_batch_request`. Would you like me to implement this here or can I do it in a follow up PR?
